pallets/loans/src/types.rs

```diff
@@ -140,18 +142,133 @@ pub enum NAVUpdateType {
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 pub struct LoanDetails<Asset, BlockNumber> {
 	pub(crate) collateral: Asset,
+
 	pub(crate) status: LoanStatus<BlockNumber>,
+
+	pub(crate) schedule: RepaymentSchedule<Moment>,
+}
+
+// TODO: implement Mid
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum InterestEvent {
+	/// At the end of the period, e.g. the last day of the month for a monthly period
+	End
+}
+
+// TODO: implement Yearly, Quarterly, Daily
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum InterestPayments {
+	/// All interest is expected to be paid at the maturity date
+	None,
+	/// Interest payments are expected monthly
+	Monthly { event: InterestEvent }
+}
+
+// TODO: implement StraightLine, Annuity
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum PayDownSchedule {
+	/// The entire borrowed amount is expected to be paid back at the maturity date
+  None
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum Maturity<Moment> {
+	Fixed(Moment),
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub struct RepaymentSchedule<Moment> {
+	/// Expected repayment date for remaining debt
+	maturity: Maturity<Moment>,
+	/// Period at which interest is paid
+	interest_payments: InterestPayments,
+	/// How much of the initially borrowed amount is paid back during interest payments
+	pay_down_schedule: PayDownSchedule
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum BorrowRestrictions {
+	None
 }
 
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum RepayRestrictions {
+	None
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum RepricingRestrictions {
+	None
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub enum MaxBorrowAmount<Rate> {
+	/// Collateral value - advance rate * total borrowed
+	UpToTotalBorrowed { advance_rate: Rate },
+	/// Collateral value - advance rate * outstanding debt
+	UpToOutstandingDebt { advance_rate: Rate },
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub struct LoanRestrictions<Rate> {
+	/// How much can be borrowed
+  max_borrow_amount: MaxBorrowAmount<Rate>,
+	/// How often can be borrowed
+  borrows: BorrowRestrictions,
+	/// How often can be repaid
+  repayments: RepayRestrictions,
+	/// How often can be priced after the initial pricing
+  repricing: RepricingRestrictions
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub struct LoanPricingInput<Rate, Balance> {
+	pub(crate) collateral_value: Balance,
+	pub(crate) interest_rate_per_year: Rate,
+	pub(crate) valuation_method: ValuationMethod<Rate, Balance>,
+	pub(crate) restrictions: LoanRestrictions<Rate>,
+}
+
+#[derive(Encode, Decode, Copy, Clone, TypeInfo)]
+pub struct LoanPricing<Rate, Balance> {
```

Comment:
Is this equals to `LoanPricingInput` because it's expected to differ in the future?

Reply:
I overlooked that! 👍🏻

Reply:
I will add a comment for it, it's clearly easy to miss ;)

Reply:
Sorry for opening this again, I was wondering if it could fit in to extract the `interest_per_second_sec` from the `LoanPricing/LoanPricingInput` and store it in the `PricedLoanDetails`. I say that because is not easy to know over which `LoanPricing` you should check things. i.e should I check if it is valid from `LoanPricing` or from `LoanPricingInput`? extracting the interest per second from it would allow only using one struct for both, as an input parameter and as a storage. 

So my real question, is. It has sense to move `interest_per_second_sec` to the `PricedLoanDetails` or semantically is it something tied to the `LoanPricing`?

Reply:
It's just that the `price` method always accepts these 4 inputs to update. So if we separate out `interest_rate_per_sec`, it just means that the `price` and possibly `reprice` method would always have the params `loan_id`, `pricing`, `interest_rate_per_sec`